[VPlan] Enable vectorization of early-exit loops with unit-stride fault-only-first loads

arcbbb opened the pull request:
Following #152422, this patch enables auto-vectorization of early-exit loops containing a single potentially faulting, unit-stride load by using the vp.load.ff intrinsic introduced in #128593.

Key changes:
- Add VPWidenFFLoadRecipe that produces two results: the loaded vector and VL (the number of lanes actually loaded).
- Use VL to ensure correctness:
   - Step the induction variable by VL (variable-length stepping).
   - Cap AVL to min(VF, remainder) to avoid over-reads at last iteration.
   - Compute the early-exit condition from VL by replacing AnyOf with vp.reduce.or to avoid branch-on-poison.
- Introduce two transforms:
   - adjustFFLoadEarlyExitForPoisonSafety: rewrites the exit condition (AnyOf → vp.reduce.or(VL)) and sets AVL = min(VF, remainder).
   - convertFFLoadEarlyExitToVLStepping: after region dissolution, converts early-exit loops to step by VL.

Limitations:
- Supports a single potentially faulting load with unit stride.
- Interleave count (IC) must be 1.

stacks on top of #165218.

arcbbb commented:
Regarding EVL propagation, I just noticed that header mask handling is fixed in #150202. This fix will need to be incorporated accordingly.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7732,6 +7741,7 @@ VPRecipeBuilder::tryToWidenMemory(Instruction *I, ArrayRef<VPValue *> Operands,
       Reverse || Decision == LoopVectorizationCostModel::CM_Widen;
 
   VPValue *Ptr = isa<LoadInst>(I) ? Operands[0] : Operands[1];
+
```
Restore

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7756,6 +7766,12 @@ VPRecipeBuilder::tryToWidenMemory(Instruction *I, ArrayRef<VPValue *> Operands,
     Ptr = VectorPtr;
   }
+  if (Legal->getSpeculativeLoads().contains(I)) {
+    LoadInst *Load = dyn_cast<LoadInst>(I);
```
```suggestion
    auto *Load = cast<LoadInst>(I);
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9985,6 +9998,25 @@ bool LoopVectorizePass::processLoop(Loop *L) {
   }
 
+  if (!LVL.getSpeculativeLoads().empty()) {
+    if (!EnableSpeculativeLoads) {
+      reportVectorizationFailure("Auto-vectorization of loops with speculative "
+                                 "load is not enabled",
+                                 "SpeculativeLoadsDisabled", ORE, L);
+      return false;
+    }
```
Check this first?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9985,6 +9998,25 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+                                 "SpeculativeLoadsDisabled", ORE, L);
+      return false;
+    }
+    // DataWithEVL is needed to lower VPWidenFFLoadRecipe into
+    // VPWidenFFLoadEVLRecipe.
```
Why?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -960,6 +967,8 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
         Instruction::Or, cast<VectorType>(VecTy), std::nullopt, Ctx.CostKind);
   }
   case VPInstruction::FirstActiveLane: {
+    if (VF.isScalar())
+      return InstructionCost::getInvalid();
```
Separate patch with a test

arcbbb commented:
> I think it would be good for reviewers if this patch was split up into several parts, probably roughly in this order:
> 
> 1. Codegen specific parts related to the RISCV intrinsic,
> 2. A standalone loop vectorisation legality patch similar to [[LV] Add initial legality checks for ee loops with stores #145663](https://github.com/llvm/llvm-project/pull/145663),
> 3. A loop vectoriser patch that starts vectorising the loops you're interested in.

Thanks for outlining the steps. That’s very helpful! I’ll follow this order for splitting up the patch.

arcbbb commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7749,6 +7758,12 @@ VPRecipeBuilder::tryToWidenMemory(Instruction *I, ArrayRef<VPValue *> Operands,
   }
+  if (Legal->getSpeculativeLoads().contains(I)) {
+    auto *Load = dyn_cast<LoadInst>(I);
+    return new VPWidenFFLoadRecipe(*Load, Ptr, Mask, VPIRMetadata(*Load, LVer),
```
Updated! After rebasing, this patch supports WidenFFLoad in non–tail-folded mode only. To tail-fold early-exit loops and FFLoad support in the EVL transform will have to be addressed in a separate patch.

arcbbb commented:
Rebased after #152422 landed, and updated the PR description as well as title.

rzinsly commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -107,6 +108,7 @@ bool VPRecipeBase::mayReadFromMemory() const {
     return cast<VPInstruction>(this)->opcodeMayReadOrWriteFromMemory();
   case VPWidenLoadEVLSC:
   case VPWidenLoadSC:
+  case VPWidenFFLoadSC:
```
Does this need to get added to `VPRecipeBase::mayHaveSideEffects()` as well?

arcbbb replied:
I lean toward treating VPWidenFFLoad as side‑effecting since it returns vector length, similar to VPInstruction::ExplicitVectorLength.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3249,6 +3258,42 @@ struct LLVM_ABI_FOR_TEST VPWidenLoadRecipe final : public VPWidenMemoryRecipe,
+/// A recipe for widening loads using fault-only-first intrinsics.
+/// Produces two results: (1) the loaded data, and (2) the index of the first
+/// non-dereferenceable lane, or VF if all lanes are successfully read.
+struct VPWidenFFLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
```
This maps one-to-one to an intrinsic, right? We support widening intrinsics with struct return types, wondering what would be needed to use VPWidenIntrinsic instead of a dedicated recipe?

arcbbb replied:
Fixed. Thanks!

arcbbb committed:
Support WidenFFLoad in early-exit loop

arcbbb committed:
[VPlan] Support struct return types for widen intrinsics (NFC)

arcbbb commented:
Rebased to include #169890.

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -399,6 +399,12 @@ static cl::opt<bool> EnableEarlyExitVectorization(
         "Enable vectorization of early exit loops with uncountable exits."));
 
+static cl::opt<bool>
+    EnableEarlyExitWithFFLoads("enable-early-exit-with-ffload", cl::init(false),
```
In the spirit of [incremental development](https://llvm.org/docs/DeveloperPolicy.html#id24) can we remove this option just have it on by default?

lukel97 reviewed (commented):
I just took a brief look at this, I'll hopefully have more time to give it a closer review next week. 

From what I understand we currently only support early exit loops without tail folding. Do we have a plan on how to support tail folding eventually and how that would interact with vp.load.ff?

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3144,6 +3144,137 @@ void VPlanTransforms::addExplicitVectorLength(
+                        m_ZeroInt(), m_VPValue(FaultingLane), m_VPValue()),
+                    m_VPValue())));
+
+  CanonicalIVIncrement->setOperand(1, FaultingLane);
```
I don't think this is going to play well with EVL tail folding because we'll now have two different transforms trying to convert the plan to variable stepping, convertFFLoadEarlyExitToVLStepping and transformRecipestoEVLRecipes.

At a high level I wonder if we even want to support vp.load.ff without EVL tail folding to begin with. This PR from what I understand is kind of reimplementing a weaker version of EVL tail folding, since the variable stepping is a hard requirement of the vp.load.ff intrinsic that we can't avoid. It can reduce the number of lanes read for any reason.

I understand that this is supposed to be an incremental PR, but I think maybe a better roadmap might be to start by supporting early exit loops with tail folding. I think this means we need to address the "variable header mask" TODO here:

```c++
bool LoopVectorizationLegality::canFoldTailByMasking() const {
  // The only loops we can vectorize without a scalar epilogue, are loops with
  // a bottom-test and a single exiting block. We'd have to handle the fact
  // that not every instruction executes on the last iteration.  This will
  // require a lane mask which varies through the vector loop body.  (TODO)
  if (TheLoop->getExitingBlock() != TheLoop->getLoopLatch()) {
    LLVM_DEBUG(
        dbgs()
        << "LV: Cannot fold tail by masking. Requires a singe latch exit\n");
    return false;
  }

```

I think we can do this if we replace the notion of a header mask with the notion of a per-block header mask.

lukel97 reviewed (commented):
I don't think this is going to play well with EVL tail folding because we'll now have two different transforms trying to convert the plan to variable stepping, convertFFLoadEarlyExitToVLStepping and transformRecipestoEVLRecipes.

At a high level I wonder if we even want to support vp.load.ff without EVL tail folding to begin with. This PR from what I understand is kind of reimplementing a weaker version of EVL tail folding, since the variable stepping is a hard requirement of the vp.load.ff intrinsic that we can't avoid. It can reduce the number of lanes read for any reason.

Trying this PR out on some llvm-test-suite benchmarks shows that the generated code always seems to generate LMUL 8 step vectors for e8 elements which is probably not great for performance:

```diff
+.LBB1197_155:                           #   Parent Loop BB1197_123 Depth=1
+                                        # =>  This Inner Loop Header: Depth=2
+       sub     a2, a0, a1
+       addi    a3, sp, 1280
+       add     a3, a3, a1
+       minu    a2, s11, a2
+       vsetvli zero, a2, e8, m2, ta, ma
+       vle8ff.v        v16, (a3)
+       csrr    a2, vl
+       csrr    a3, vlenb
+       vsetvli a4, zero, e64, m8, ta, ma
+       vid.v   v8
+       vmv.v.v v24, v8
+       vadd.vx v8, v8, a3
+       zext.w  a4, a2
+       vmsltu.vx       v18, v8, a4
+       vmsltu.vx       v8, v24, a4
+       srli    a4, a3, 2
+       vsetvli a5, zero, e8, m2, ta, ma
+       vmseq.vi        v9, v16, 0
+       srli    a3, a3, 3
+       vsetvli zero, a4, e8, mf4, ta, ma
+       vslideup.vx     v8, v18, a3
+       vsetvli a3, zero, e8, m2, ta, ma
+       vmand.mm        v8, v8, v9
+       vcpop.m a3, v8
+       bnez    a3, .LBB1197_157
+# %bb.156:                              #   in Loop: Header=BB1197_155 Depth=2
+       add.uw  a1, a2, a1
+       bne     a1, a0, .LBB1197_155
+.LBB1197_157:                           #   in Loop: Header=BB1197_123 Depth=1
+       snez    a1, a3
+       beqz    a1, .LBB1197_159
+.LBB1197_158:                           #   in Loop: Header=BB1197_123 Depth=1
```

So I don't think there's really much reason why we would want to emit non-tail folded early-exit loops if we can tail fold them eventually.

I understand that this is supposed to be an incremental PR, but I think maybe a better ordering might be to start by supporting early exit loops with tail folding. I think this means we need to address the "variable header mask" TODO here:

```c++
bool LoopVectorizationLegality::canFoldTailByMasking() const {
  // The only loops we can vectorize without a scalar epilogue, are loops with
  // a bottom-test and a single exiting block. We'd have to handle the fact
  // that not every instruction executes on the last iteration.  This will
  // require a lane mask which varies through the vector loop body.  (TODO)
  if (TheLoop->getExitingBlock() != TheLoop->getLoopLatch()) {
    LLVM_DEBUG(
        dbgs()
        << "LV: Cannot fold tail by masking. Requires a singe latch exit\n");
    return false;
  }

```

I think we can do this if we replace the notion of a header mask with the notion of a per-block header mask. I'll see if I can create an issue to discuss some of the design of this more.

arcbbb committed:
Reuse getCostForIntrinsics and remove special handling for vp_load_ff

arcbbb commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3144,6 +3144,137 @@ void VPlanTransforms::addExplicitVectorLength(
+    return;
+
+  // Ensure FFLoad does not read past the remainder in the last iteration.
+  // Set AVL to min(VF, remainder).
```
This is primarily a correctness issue. Since the header mask is derived from the second returned value of vp.load.ff, we can use it directly only if the AVL is capped before the call.
For example:
```
  %9 = call { <vscale x 16 x i8>, i32 } @llvm.vp.load.ff.nxv16i8.p0(ptr %8, <vscale x 16 x i1> splat (i1 true), i32 %7)
  %10 = extractvalue { <vscale x 16 x i8>, i32 } %9, 1
  %11 = zext i32 %10 to i64
  %14 = call <vscale x 16 x i1> @llvm.get.active.lane.mask.nxv16i1.i64(i64 0, i64 %11)
```
get.active.lane.mask generates the predicate mask and uses %11 directly because AVL was capped earlier. If not, we would cap it before calling get.active.lane.mask.

david-arm commented on llvm/test/Transforms/LoopVectorize/RISCV/find.ll:
I think we need tests in the top level LoopVectorize directory too. For example, I think it's worth adding an extra RUN line with the -enable-early-exit-with-ffload flag to Transforms/LoopVectorize/single_early_exit.ll and Transforms/LoopVectorize/single_early_exit_live_outs.ll.

Those tests have a very extensive coverage of different CFGs and combinations of live-outs in different scenarios.

arcbbb committed:
Address comments

- Merge `canonicalizeEVLLoops` and `convertFFLoadEarlyExitToVLStepping`
  into a single entry point `convertLoopIntoVariableLengthStepping`. Both
  transforms convert loops to use variable-length stepping after region
  dissolution.

- Generalize poison-safety handling in `adjustFFLoadEarlyExitForPoisonSafety`
  to mask all AnyOf users of FFLoad (not just a single hardcoded pattern).

arcbbb committed:
Rebase code

lukel97 reviewed (commented):
I've opened up some thoughts on fault-only-first loads in #175900, but thought I'd mention some things on this PR that I'd like to hear your thoughts on:

1. fault-only-first loads require variable stepping, which you've converted in this PR. But I think we still need to update induction recipes to account for the fact that they no longer step by VF?
2. `@llvm.experimental.get.vector.length` has the property that the step amount may vary on the penultimate iteration, but otherwise the vector loop will always be executed the same number of iterations as a non tail folded loop. i.e. the backedge taken count is always the same.
I don't think this is true of `@llvm.vp.load.ff`. It's valid for a vle*ff.v to reduce VL to one for every iteration, which means the vector trip count may change. Do we need to go through any users of the vector trip count and check if they need adapted?
3. We don't predicate any recipes after the `vp.load.ff`. I think this happens to work at the moment because we currently don't allow any non-speculatable recipes in early-exit loops, but this won't be the case after #172454.
I think it would be good to handle this logic together in this PR, otherwise it will surprisingly break when more non-speculatable recipes are allowed. I think landing #172454 first would allow us to write test cases for this.

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
I presume this needs to be rebased to use VPCurrentIterationPHI now?

arcbbb replied:
Yes! I plan to rebase and rework this patch after https://github.com/llvm/llvm-project/pull/180039

lukel97 commented on llvm/lib/Analysis/VectorUtils.cpp:
Are these needed for this PR?
